### PR TITLE
JIRA Integration

### PR DESCRIPTION
**Pull request by djcooke**

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditProjectController.java [hunk 1]
```diff
@@ -259,18 +267,25 @@ public ModelAndView byProjectShortName(@PathVariable String shortName, ModelMap
...
       if (projectId == ProjectImpl.UNSAVED_ID) {
         project = new ProjectImpl(user);
         model.put("title", "New Project");
       } else {
         project = projectService.getProjectById(projectId);
+        if (project == null) {
+          throw new NotFoundException("No project found for ID " + projectId.toString());
+        }
         model.put("title", "Project " + projectId);
+        try {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditProjectController.java]: MAJOR Extract this nested try block into a separate method. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditProjectController.java [hunk 2]
```diff
@@ -259,18 +267,25 @@ public ModelAndView byProjectShortName(@PathVariable String shortName, ModelMap
   @RequestMapping(value = "/{projectId}", method = RequestMethod.GET)
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditProjectController.java]: MINOR Replace "@RequestMapping(method = RequestMethod.GET)" with "@GetMapping" [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4488)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 3]
```diff
@@ -58,24 +67,39 @@
...
+  private URI getRestUri(String relativeUrl, Map<String, String> params) {
+    StringBuilder sb = new StringBuilder();
+    sb.append(baseTrackerUrl).append(restApiUrl).append(jiraRestApiVersion);
+    if (!relativeUrl.startsWith("/")) {
+      sb.append("/");
+    }
+    sb.append(relativeUrl);
+    if (params != null && !params.isEmpty()) {
+      sb.append("?");
+      params.forEach((key, value) -> {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: MINOR Remove useless curly braces around statement [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1602)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 4]
```diff
@@ -58,24 +67,39 @@
...
   private String httpBasicAuthUsername;
   private String httpBasicAuthPassword;
 
-  public String baseTrackerUrl;
+  private String baseTrackerUrl;
   private final String restApiUrl = "/rest/api/";
-  public String jiraRestApiVersion = "2";
-
-  public final String jiraIssueSuffix = restApiUrl + jiraRestApiVersion + "/issue/";
+  private final String jiraRestApiVersion = "2";
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: MINOR Make this final field static too. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1170)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 5]
```diff
@@ -58,24 +67,39 @@
  */
 @Component
 public class JiraIssueManager implements IssueTrackerManager {
+
+  private static final DateFormat ISO8601_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: MAJOR Make "ISO8601_FORMAT" an instance variable. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2885)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 6]
```diff
@@ -58,24 +67,39 @@
...
+  private static final List<Character> SPECIAL_CHARS = Lists.newArrayList('+', '-', '&', '|', '!', '(', ')', '{', '}', '[', ']', '^', '~',
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: MAJOR Remove this unused "SPECIAL_CHARS" private field. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1068)

**Comment by sqwci** [comment]
SonarQube analysis reported 33 issues
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 10 major
* ![MINOR][MINOR] 20 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [Project.java#L47](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/Project.java#L47): Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)
1. ![CRITICAL][CRITICAL] [EditRunController.java#L211](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditRunController.java#L211): Define a constant instead of duplicating this literal "partitionNames" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditRunController.java#L287](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditRunController.java#L287): Replace this persistent entity with a simple POJO or DTO object. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4684)
1. ![MAJOR][MAJOR] [DefaultProjectService.java#L165](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L165): Return an empty collection instead of null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [DefaultProjectService.java#L177](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L177): Return an empty collection instead of null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [MisoAppListener.java#L157](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/MisoAppListener.java#L157): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [MisoAppListener.java#L219](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/MisoAppListener.java#L219): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [EditRunController.java#L216](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditRunController.java#L216): Refactor the code so this stream pipeline is used. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3958)
1. ![MAJOR][MAJOR] [ProjectControllerHelperService.java#L124](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/spring/src/main/java/uk/ac/bbsrc/tgac/miso/spring/ajax/ProjectControllerHelperService.java#L124): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [ProjectControllerHelperService.java#L472](https://github.com/TGAC/miso-lims/blob/3ed95b0ceca679c86ee8a07cbbc197ae237df50b/spring/src/main/java/uk/ac/bbsrc/tgac/miso/spring/ajax/ProjectControllerHelperService.java#L472): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 7]
```diff
@@ -58,55 +66,52 @@
  */
 @Component
 public class JiraIssueManager implements IssueTrackerManager {
+
+  private static final String restApiUrl = "/rest/api/";
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 8]
```diff
@@ -58,55 +66,52 @@
...
+  private static final String jiraRestApiVersion = "2";
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: CRITICAL Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00115)

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java [hunk 9]
```diff
@@ -58,55 +66,52 @@
...
+
+  private final DateFormat ISO8601_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ");
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/service/integration/jira/JiraIssueManager.java]: MINOR Rename this field "ISO8601_FORMAT" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00116)

**Comment by sqwci** [comment]
SonarQube analysis reported 26 issues
* ![CRITICAL][CRITICAL] 5 critical
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 14 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [Project.java#L47](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/Project.java#L47): Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)
1. ![CRITICAL][CRITICAL] [EditRunController.java#L211](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditRunController.java#L211): Define a constant instead of duplicating this literal "partitionNames" 3 times. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)
1. ![CRITICAL][CRITICAL] [EditRunController.java#L287](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditRunController.java#L287): Replace this persistent entity with a simple POJO or DTO object. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4684)
1. ![MAJOR][MAJOR] [DefaultProjectService.java#L165](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L165): Return an empty collection instead of null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [DefaultProjectService.java#L177](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultProjectService.java#L177): Return an empty collection instead of null. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1168)
1. ![MAJOR][MAJOR] [MisoAppListener.java#L157](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/MisoAppListener.java#L157): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [MisoAppListener.java#L219](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/MisoAppListener.java#L219): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [EditRunController.java#L216](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditRunController.java#L216): Refactor the code so this stream pipeline is used. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3958)
1. ![MAJOR][MAJOR] [ProjectControllerHelperService.java#L124](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/spring/src/main/java/uk/ac/bbsrc/tgac/miso/spring/ajax/ProjectControllerHelperService.java#L124): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)
1. ![MAJOR][MAJOR] [ProjectControllerHelperService.java#L472](https://github.com/TGAC/miso-lims/blob/484665474d6c2c55d6958ba6ffa3d88c49699253/spring/src/main/java/uk/ac/bbsrc/tgac/miso/spring/ajax/ProjectControllerHelperService.java#L472): Format specifiers should be used instead of string concatenation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3457)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/MisoAppListener.java [hunk 10]
```diff
@@ -127,22 +126,17 @@ public void contextInitialized(ServletContextEvent event) {
...
-        }
+      if ("jira".equals(trackerType)) {
+        IssueTrackerManager issueTracker = new JiraIssueManager();
+        Properties properties = new Properties();
+        properties.putAll(misoProperties);
+        issueTracker.setConfiguration(properties);
+        ((DefaultListableBeanFactory) context.getBeanFactory()).removeBeanDefinition("issueTrackerManager");
+        context.getBeanFactory().registerSingleton("issueTrackerManager", issueTracker);
+      } else {
+        throw new IllegalArgumentException("Invalid tracker type specified at miso.issuetracker.tracker");
```
[the remark this answers is not in the document]
Comment (djcooke) [reply miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/MisoAppListener.java]: Depending on the value, 'invalid' may make more sense than 'unsupported.' The message will only show up in logs though, and I think gets the point across either way

**Pull request merged**